util/src/main/java/io/zeebe/util/health/CriticalComponentsHealthMonitor.java

```diff
@@ -0,0 +1,120 @@
+  private static final Duration HEALTH_MONITORING_PERIOD = Duration.ofSeconds(60);
+  private final Map<String, HealthMonitorable> monitoredComponents = new HashMap<>();
+  private final Map<String, HealthStatus> componentHealth = new HashMap<>();
+  private final AtomicReference<HealthStatus> healthStatus = new AtomicReference<>();
+  private final ActorControl actor;
+  private final Logger log;
+  private FailureListener failureListener;
+
+  public CriticalComponentsHealthMonitor(final ActorControl actor, final Logger log) {
+    this.actor = actor;
+    this.log = log;
+    this.failureListener = null;
+  }
+
+  public void startMonitoring() {
+    actor.runAtFixedRate(HEALTH_MONITORING_PERIOD, this::updateHealth);
+  }
+
+  @Override
+  public void monitorComponent(final String componentName) {
+    actor.run(() -> componentHealth.put(componentName, HealthStatus.UNHEALTHY));
+  }
+
+  @Override
+  public void removeComponent(final String componentName) {
+    actor.run(
+        () -> {
+          monitoredComponents.remove(componentName);
+          componentHealth.remove(componentName);
+        });
+  }
+
+  @Override
+  public void registerComponent(final String componentName, final HealthMonitorable component) {
+    actor.run(
+        () -> {
+          monitoredComponents.put(componentName, component);
+          component.addFailureListener(
+              FailureListener.withListeners(
+                  () -> onComponentFailure(componentName),
+                  () -> onComponentRecovered(componentName)));
+          componentHealth.put(componentName, component.getHealthStatus());
+        });
+  }
+
+  @Override
+  public HealthStatus getHealthStatus() {
+    return healthStatus.get();
+  }
+
+  @Override
+  public void addFailureListener(final FailureListener failureListener) {
+    actor.run(() -> this.failureListener = failureListener);
+  }
+
+  private void onComponentFailure(final String componentName) {
+    actor.call(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+          reCalculateHealth();
+        });
+  }
+
+  private void onComponentRecovered(final String componentName) {
+    actor.call(
+        () -> {
+          log.error("{} failed, marking it as unhealthy", componentName);
+          componentHealth.put(componentName, HealthStatus.UNHEALTHY);
+        });
+  }
+
+  private void updateHealth() {
+    componentHealth.keySet().stream()
+        .forEach(component -> componentHealth.put(component, getHealth(component)));
+    reCalculateHealth();
+  }
+
+  private void reCalculateHealth() {
+    final boolean healthy =
+        componentHealth.values().stream()
+            .allMatch(healthStatus -> healthStatus == HealthStatus.HEALTHY);
+    healthStatus.set(healthy ? HealthStatus.HEALTHY : HealthStatus.UNHEALTHY);
+    if (!healthy) {
```

Comment:
Should we notify the listener if we go from `UNHEALTHY` to `UNHEALTHY` again? Not saying we shouldn't, I'm simply not sure what we want to do - notify of every new failure, notify only of status changes?

Should we notify that we recovered?

Reply:
You are right. We should notify when the status changes.